### DISCUSSION TITLE
Load Tester not generating requests/tunnels as per the configuration

### DISCUSSION THREAD

**Discussion by dhanaseelangit** [strongswan/strongswan#1824] [Q&A]
Hi All,

Sorry for the long message. I am working on one of the test case to load the real IPSec server with tunnels and When I explore the Strongswan I came to know that the load-tester plugin helps to load the Server. I started using the load-tester and found below observation and need suggestions to correct the behaviour of my IPSec client.

I am using the **Ubuntu(22.04) based docker environment for my testing** and I compiled the **source code of strongswan version 5.9.11**.

* Used the **Load tester with configuration as initiators as 30, iterations as 1000 and delay as 100 to generate 30K tunnels with 300 taunnels per second**. In total as per the configuration 30K tunnel request should be generated in ~30 seconds.
* At the start of the testing I could see the tunnel request are sent almost 300 per second, but after sometime its started degrading and **it took around 7 minutes to reach ~9.5K tunnels.**
* I confirmed the **number of request reaching the IPSec server is going down from nearly 288+ to 12+ over time**. I confirmed the same with the stats at IPSec server.
* Throughout the testing I could see only **max 20% of the CPU is utilized and around 1/10th of the memory consumed.**
* Post 9.5K sessions the charon is crashed with segmentation fault.
* Tried With multiple runs of the same test case. **I could not able to reach max tunnels of 30K tunnels** in any of the runs.
* Attached the “swanctl —stats” output collected at IPSec client side at different intervals.

**Is there any similar issues seen by anyone in the past. If seen, can you please suggest the way to resolve or correct?**

root@cb870b0e54c1:/# swanctl --version
strongSwan swanctl 5.9.11
root@cb870b0e54c1:/#

**Load-tester.conf**

    addrs {

    }

    # enable the plugin
    enable = yes
    # 10000 connections, ten in parallel
    initiators = 30
    iterations = 1000
    # use a delay of 100ms, overall time is: iterations * delay = 100s
    delay = 100
    # address of the gateway (releases before 5.0.2 used the "remote" keyword!)
    responder = *********
    # IKE-proposal to use
    proposal = aes256-sha256-modp2048
    # use faster PSK authentication instead of 1024bit RSA
    initiator_auth = psk
    responder_auth = psk
    # request a virtual IP using configuration payloads
    request_virtual_ip = yes
    # disable IKE_SA rekeying (default)
    ike_rekey = 0
    # enable CHILD_SA every 60s
    child_rekey =600
    # do not delete the IKE_SA after it has been established (default)
    delete_after_established = no
    # do not shut down the daemon if all IKE_SAs established
    shutdown_when_complete = no

    initiator_tsr = ::/0
    initiator_tsi = ::/0
    issuer_cert = /etc/swanctl/x509ca
    initiator_id = BG0000006%d-%d
    esp = aes256-sha1
    preshared_key =********

**Strongswan.conf**

charon {
        load_modular = yes
        threads = 3100
        plugins {
                include strongswan.d/charon/*.conf
        }
        filelog {
                charon {
                        path = /var/log/charon.log
                        ike_name = yes
                        time_format = %b %e %T
                        append = no
                        default = 2
                }
        }
        syslog {
                auth {
                        default = -1
                }
        }
        processor {
                priority_threads {
                        high = 5
                        #medium = 500
                }
        }

}

include strongswan.d/*.conf

Below are the outputs collected during one of the run and could see request generation per seconds is degrading over the time.

root@cb870b0e54c1:/# swanctl --stats
uptime: **61 seconds**, since Aug 01 15:17:06 2023
worker threads: 3100 total, 2911 idle, working: 34/0/155/0
job queues: 0/0/0/0
jobs scheduled: 2275
**IKE_SAs: 2224 total, 6 half-open**
mallinfo: sbrk 170708992, mmap 135168, used 162017504, free 8691488
loaded plugins: charon random nonce x509 constraints pubkey pem openssl load-tester kernel-netlink resolve socket-default vici updown eap-identity eap-md5 eap-dynamic eap-tls counters

root@cb870b0e54c1:/# swanctl --stats
**uptime: 94 seconds**, since Aug 01 15:17:06 2023
worker threads: 3100 total, 2885 idle, working: 34/4/175/2
job queues: 0/0/0/0
jobs scheduled: 2429
**IKE_SAs: 2392 total, 4 half-open**
mallinfo: sbrk 188063744, mmap 135168, used 173811744, free 14252000
loaded plugins: charon random nonce x509 constraints pubkey pem openssl load-tester kernel-netlink resolve socket-default vici updown eap-identity eap-md5 eap-dynamic eap-tls counters

root@cb870b0e54c1:/# swanctl --stats
**uptime: 118 seconds,** since Aug 01 15:17:06 2023
worker threads: 3100 total, 2880 idle, working: 34/1/185/0
job queues: 0/0/0/0
jobs scheduled: 2535
**IKE_SAs: 2501 total, 4 half-open**
mallinfo: sbrk 200032256, mmap 135168, used 182464640, free 17567616
loaded plugins: charon random nonce x509 constraints pubkey pem openssl load-tester kernel-netlink resolve socket-default vici updown eap-identity eap-md5 eap-dynamic eap-tls counters

root@cb870b0e54c1:/# swanctl --stats
**uptime: 7 minutes**, since Aug 01 15:17:06 2023
worker threads: 3100 total, 3037 idle, working: 34/8/21/0
job queues: 0/0/0/0
jobs scheduled: 9603
**IKE_SAs: 9443 total, 10 half-open**
mallinfo: sbrk 407683072, mmap 1187840, used 403753328, free 3929744
loaded plugins: charon random nonce x509 constraints pubkey pem openssl load-tester kernel-netlink resolve socket-default vici updown eap-identity eap-md5 eap-dynamic eap-tls counters

**Few of the crashes seen during the run,**

root@cb870b0e54c1:/# ./charon 
addr2line: 'linux-vdso.so.1': No such file
 dumping 10 stack frame addresses:
  linux-vdso.so.1 @ 0xffffae37c000 (__kernel_rt_sigreturn+0x0) [0xffffae37c7c0]
addr2line: 'linux-vdso.so.1': No such file
    -> 
  /usr/lib/ipsec/libcharon.so.0 @ 0xffffae210000 [0xffffae2689c8]
    -> /strongswan-build/strongswan-5.9.11/src/libcharon/sa/ikev2/tasks/ike_config.c:502
  /usr/lib/ipsec/libcharon.so.0 @ 0xffffae210000 [0xffffae25da58]
    -> /strongswan-build/strongswan-5.9.11/src/libcharon/sa/ikev2/tasks/child_create.c:2059
  /usr/lib/ipsec/libcharon.so.0 @ 0xffffae210000 [0xffffae258958]
    -> /strongswan-build/strongswan-5.9.11/src/libcharon/sa/ikev2/task_manager_v2.c:839
  /usr/lib/ipsec/libcharon.so.0 @ 0xffffae210000 [0xffffae244eb0]
    -> /strongswan-build/strongswan-5.9.11/src/libcharon/sa/ike_sa.c:1636
  /usr/lib/ipsec/libcharon.so.0 @ 0xffffae210000 [0xffffae23dc58]
    -> /strongswan-build/strongswan-5.9.11/src/libcharon/processing/jobs/process_message_job.c:75
  /usr/lib/ipsec/libstrongswan.so.0 @ 0xffffae2b0000 [0xffffae2f0b5c]
    -> /strongswan-build/strongswan-5.9.11/src/libstrongswan/processing/processor.c:262
  /usr/lib/ipsec/libstrongswan.so.0 @ 0xffffae2b0000 [0xffffae304574]
    -> /strongswan-build/strongswan-5.9.11/src/libstrongswan/threading/thread.c:336 (discriminator 4)
  /lib/aarch64-linux-gnu/libc.so.6 @ 0xffffae060000 [0xffffae0dd5c8]
    -> ??:?
  /lib/aarch64-linux-gnu/libc.so.6 @ 0xffffae060000 [0xffffae145d1c]
    -> ??:?
Aborted
root@cb870b0e54c1:/# ./c 

Please let me know if any other details would be of help to isolate the problem and correct the behaviour.

Thanks,
Dhana

**Comment by tobiasbrunner** [comment]
> threads = 3100

That definitely won't help. How many cores does your host have? More than twice that number will probably not be helpful (unless there are blocking calls like RADIUS or DHCP, which isn't the case here).

You probably want to enable the [IKE_SA hash table to improve lookups](https://docs.strongswan.org/docs/5.9/config/lookupTuning.html) when there are many IKE_SAs. Or you could enable `delete_after_established` to keep the number of established SAs low.

Also be aware that much of the slow down could be caused by the responder (dropped packets, retransmits etc.).

> **Few of the crashes seen during the run,**

That looks weird. Not only do the line numbers not match (at least when I look at the code here, maybe you have some patches applied). There is also no code path at all from `child_create.c::destroy()` (around where the line number is) to `ike_config.c::destroy()` as the child-create task never holds a reference to an ike-config task. So no idea what's going on there.

**Reply by dhanaseelangit** [reply]
Hi Tobiasbrunner,

Thanks a lot with the changes of ikesa_table_size and ikesa_table_segments, I don't see any crashes. I had 8 cores on my CPU. So setting the threads to 8.

charon {
        load_modular = yes
        threads = 8
       ikesa_table_size = 2048
       ikesa_table_segments = 16
       reuse_ikesa = no
       keep_alive = 6000
        plugins {
                include strongswan.d/charon/*.conf
        }
        filelog {
                charon {
                        path = /var/log/charon.log
                        ike_name = yes
                        time_format = %b %e %T
                        append = no
                        default = 2
                }
        }
        syslog {
                auth {
                        default = -1
                }
        }
        processor {
                priority_threads {
                        #high = 1
                        #medium = 500
                }
        }

}

include strongswan.d/*.conf

With the above config I could able to generate 30K IKE_SA Init req, got the IKE_SA_Init Response from IPSec server. 

IKE Auth Req is generated for all the Tunnels but around 1400 IKE responses only received on the charon. I tried capturing the PCAP and I could around 8000 Auth responses received. At the same time, charon process is around 54% loaded. I could see only 1400 tunnels established successfully at the end.

Any suggestions to improve the tunnels success rate.

Thanks,
Dhana

**Reply by tobiasbrunner** [reply]
8 threads is a bit low as some threads will be reserved for long-running tasks (e.g. listening on various sockets, depends on the plugins etc.). There should be at least as many free worker threads as you have CPU cores right after startup when the daemon is idle (check the stats output).

> IKE Auth Req is generated for all the Tunnels but around 1400 IKE responses only received on the charon. I tried capturing the PCAP and I could around 8000 Auth responses received.

How do you get to the number of responses received by charon? What happens to the other packets? Or are some of the others retransmitted responses?

Also, are retransmits of IKE_AUTH requests sent for the other SAs? What happens on the server side?

**Comment by dhanaseelangit** [comment]
Anyone has any suggestions??
